core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+					)?;
+
+					let (_, inner_body) = block.deconstruct();
+					body = Some(inner_body);
+				}
+
+				trace!(target: "babe", "Checked {:?}; importing.", pre_header);
+				telemetry!(CONSENSUS_TRACE; "babe.checked_and_importing"; "pre_header" => ?pre_header);
+
+				extra_verification.into_future().wait()?;
+
+				let import_block = ImportBlock {
+					origin,
+					header: pre_header,
+					post_digests: vec![seal],
+					body,
+					finalized: false,
+					justification,
+					auxiliary: Vec::new(),
+					fork_choice: ForkChoiceStrategy::LongestChain,
+				};
+
+				// FIXME #1019 extract authorities
+				Ok((import_block, None))
+			}
+			CheckedHeader::Deferred(a, b) => {
+				debug!(target: "babe", "Checking {:?} failed; {:?}, {:?}.", hash, a, b);
+				telemetry!(CONSENSUS_DEBUG; "babe.header_too_far_in_future";
+					"hash" => ?hash, "a" => ?a, "b" => ?b
+				);
+				Err(format!("Header {:?} rejected: too far in the future", hash))
+			}
+		}
+	}
+}
+
+impl<B, C, E> Authorities<B> for BabeVerifier<C, E> where
+	B: Block,
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+{
+	type Error = ConsensusError;
+
+	fn authorities(&self, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, Self::Error> {
+		authorities(self.client.as_ref(), at)
+	}
+}
+
+fn authorities<B, C>(client: &C, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, ConsensusError> where
+	B: Block,
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+{
+	client
+		.cache()
+		.and_then(|cache| cache.get_at(&well_known_cache_keys::AUTHORITIES, at)
+			.and_then(|v| Decode::decode(&mut &v[..])))
+		.or_else(|| {
+			if client.runtime_api().has_api::<AuthoritiesApi<B>>(at).unwrap_or(false) {
+				AuthoritiesApi::authorities(&*client.runtime_api(), at).ok()
+			} else {
+				panic!("We don’t support deprecated code with new consensus algorithms, therefore this is unreachable; qed")
+			}
+		}).ok_or_else(|| consensus_common::ErrorKind::InvalidAuthoritiesSet.into())
+}
+
+/// The BABE import queue type.
+pub type BabeImportQueue<B> = BasicQueue<B>;
+
+/// Register the babe inherent data provider, if not registered already.
+fn register_babe_inherent_data_provider(
+	inherent_data_providers: &InherentDataProviders,
+	slot_duration: u64,
+) -> Result<(), consensus_common::Error> {
+	debug!(target: "babe", "Registering");
+	if !inherent_data_providers.has_provider(&srml_babe::INHERENT_IDENTIFIER) {
+		inherent_data_providers
+			.register_provider(srml_babe::InherentDataProvider::new(slot_duration))
+			.map_err(inherent_to_common_error)
+	} else {
+		Ok(())
+	}
+}
+
+fn get_keypair(q: &sr25519::Pair) -> &Keypair {
+	q.as_ref()
+}
+
+/// Pieces:
+///
+/// 1. VRF winnership
+/// 2. * VRF randomness
+///    * relative time
+
+fn make_transcript(
+	randomness: &[u8],
+	slot_number: u64,
+	genesis_hash: &[u8],
+	epoch: u64,
+) -> Transcript {
+	let mut transcript = Transcript::new(&BABE_ENGINE_ID);
+	transcript.proto_name(&BABE_ENGINE_ID);
```

Comment:
You do not have to call proto_name here.  You've already committed it to the transcript in new.  The idea of proto_name is that when you pass a transcript into something else, then the something else identifies itself by calling proto_name.  It saves you from thinking about whether the contexts on some of the commit_bytes commands differ.